runtime/memprof.c

```diff
@@ -206,1126 +1258,921 @@ void caml_memprof_update_suspended(bool s) {
+    res = entries_run_callbacks_exn(thread, &domain->entries);
+    if (Is_exception_result(res)) goto end;
 
-static void entry_array_minor_update(struct entry_array *ea, void *data)
-{
-  uintnat i;
-  (void)data;
-  /* See comment in [entry_array_oldify_young_roots] for the number
-     of iterations of this loop. */
-  for (i = ea->young_idx; i < ea->len; i++) {
-    struct tracked *t = &ea->t[i];
-    CAMLassert(Is_block(t->block) || t->deleted || t->deallocated ||
-               Is_placeholder(t->block));
-    if (Is_block(t->block) && Is_young(t->block)) {
-      if (Hd_val(t->block) == 0) {
-        /* Block has been promoted */
-        t->block = Field(t->block, 0);
-        t->promoted = 1;
-      } else {
-        /* Block is dead */
-        CAMLassert_young_header(Hd_val(t->block));
-        t->block = Val_unit;
-        t->deallocated = 1;
-      }
+    /* run per-thread callbacks for current thread */
+    res = entries_run_callbacks_exn(thread, &thread->entries);
+    if (Is_exception_result(res)) goto end;
+    /* Move any surviving entries from allocating thread to owning
+     * domain, so their subsequent callbacks may be run by any thread in
+     * the domain. */
+    entries_transfer(&thread->entries, &domain->entries);
+
+    memprof_orphan_table_t ot = domain->orphans;
+    while (ot) {
+      res = entries_run_callbacks_exn(thread, &ot->entries);
+      if (Is_exception_result(res))
+        break;
+
+      ot = ot->next;
     }
-  }
-  ea->young_idx = ea->len;
-}
 
-void caml_memprof_minor_update(void)
-{
-  if (callback_idx > entries_global.young_idx) {
-    /* The entries after [entries_global.young_idx] will possibly get
-       promoted. Hence, there might be pending promotion callbacks. */
-    callback_idx = entries_global.young_idx;
-    check_action_pending();
+  end:
+    orphans_update_pending(domain);
+    update_suspended(domain, false);
   }
 
-  entry_arrays_iter(entry_array_minor_update, NULL);
+  return res;
 }
 
-static void entry_array_do_roots(struct entry_array *ea, void* data)
-{
-  scanning_action f = data;
-  uintnat i;
-  for (i = 0; i < ea->len; i++)
-    f(ea->t[i].user_data, &ea->t[i].user_data);
-}
+/**** Sampling ****/
 
-void caml_memprof_do_roots(scanning_action f)
+Caml_inline bool sampling(memprof_domain_t domain)
 {
-  entry_arrays_iter(entry_array_do_roots, f);
-}
+  memprof_thread_t thread = domain->current;
 
-static void entry_array_clean_phase(struct entry_array *ea, void* data)
-{
-  uintnat i;
-  (void)data;
-  for (i = 0; i < ea->len; i++) {
-    struct tracked *t = &ea->t[i];
-    if (Is_block(t->block) && !Is_young(t->block)) {
-      CAMLassert(Is_in_heap(t->block));
-      CAMLassert(!t->alloc_young || t->promoted);
-      if (Is_white_val(t->block)) {
-        t->block = Val_unit;
-        t->deallocated = 1;
-      }
-    }
+  if (thread && !thread->suspended) {
+    value config = thread_config(thread);
+    return Sampling(config) && Lambda(config) > 0.0;
```

Comment:
Is the second condition needed? Shouldn't it just be an error to sample with lambda=0?

Reply:
Maybe. It was allowed in OCaml 4, and there are tests in the test suite which use it, so I allowed it. Some future refinement might take it out.

Reply:
I infer that this is a concern with very small but non-zero lambda? If zero lambda is allowed (as at present, and tested in the test suite) then `caml_memprof_start` could set lambda to zero if `one_log1m_lambda` is negative infinity. Then the various tests, here and in `rand_init`, `rand_geom`, etc, would continue to work. Experimentally, we get negative infinity if lambda < 2^-1023, at which point IMO it's reasonable to clamp it to zero and never sample anything.

Reply:
I want to continue storing `lambda` for debugging and in the expectation that some future extension will want to obtain it. I am adding a change to spot when `one_log1m_lambda` is infinite (of either sign) and setting `lambda` to zero and `one_log1m_lambda` to negative infinity in that case.

Reply:
Reflecting, I'd rather retain the user's value of `lambda` in this case, for debugging. So I'll just clamp `one_log1m_lambda` to negative infinity in that case, and test against it.